lib/resty/rediscluster.lua

```diff
@@ -128,28 +138,40 @@ local function try_hosts_slots(self, serv_list)
 
         --attempt to connect DEFAULT_MAX_CONNECTION_ATTEMPTS times to redis
         for k = 1, config.max_connection_attempts or DEFAULT_MAX_CONNECTION_ATTEMPTS do
+            ok, err = redis_client:connect(ip, port, self.config.connect_opts)
+            if ok then break end
+            if err then
+                ngx.log(ngx.ERR, "unable to connect, attempt nr ", k, " : error: ", err)
+                table_insert(errors, err)
+            end
+
             local total_connection_time_ms = (ngx.now() - start_time) * 1000
             if (config.max_connection_timeout and total_connection_time_ms > config.max_connection_timeout) then
                 max_connection_timeout_err = "max_connection_timeout of " .. config.max_connection_timeout .. "ms reached."
                 ngx.log(ngx.ERR, max_connection_timeout_err)
                 table_insert(errors, max_connection_timeout_err)
                 break
             end
-
-            ok, err = redis_client:connect(ip, port, self.config.connect_opts)
-            if ok then break end
-            if err then
-                ngx.log(ngx.ERR,"unable to connect, attempt nr ", k, " : error: ", err)
-                table_insert(errors, err)
-            end
         end
 
         if ok then
             local _, autherr = check_auth(self, redis_client)
             if autherr then
+                redis_client:close()
                 table_insert(errors, autherr)
                 return nil, errors
             end
+            local info_server, info_err = redis_client:info("server")
+            if type(info_server) ~= "string" then
+                redis_client:close()
+                table_insert(errors, info_err)
+                return nil, errors
+
+            else
+                -- retrieve major and minor only
+                redis_version = tonumber(info_server:match("redis_version:(%d+%.%d+)"))
+            end
+
             local slots_info
             slots_info, err = redis_client:cluster("slots")
```

Comment:
First parse the slots_info into an easy to use table, then do the processing that is required for node selection on that table.  Right now, the use of the raw data makes it impossible to quickly understand the code and to verify that it is doing the right thing.

Reply:
Updated the code with samples and code refactor.